frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts

```diff
@@ -0,0 +1,422 @@
+    this.rescueEditorContent();
+    this.removeEventListeners();
+    this.stopPolling();
+  }
+
+  private setLocalStorageKey() {
+    // scoped by user id in order to avoid data leakage when a user logs out and another user logs in on the same browser
+    // TODO: when a user logs out, the data should be removed anyways in order to avoid data leakage
+    this.localStorageKey = `work-package-${this.workPackageIdValue}-rescued-editor-data-${this.userIdValue}`;
+  }
+
+  private setupEventListeners() {
+    this.handleWorkPackageUpdateBound = this.handleWorkPackageUpdate.bind(this);
+    this.handleVisibilityChangeBound = this.handleVisibilityChange.bind(this);
+    this.rescueEditorContentBound = this.rescueEditorContent.bind(this);
+
+    document.addEventListener('work-package-updated', this.handleWorkPackageUpdateBound);
+    document.addEventListener('visibilitychange', this.handleVisibilityChangeBound);
+    document.addEventListener('beforeunload', this.rescueEditorContentBound);
+  }
+
+  private removeEventListeners() {
+    document.removeEventListener('work-package-updated', this.handleWorkPackageUpdateBound);
+    document.removeEventListener('visibilitychange', this.handleVisibilityChangeBound);
+    document.removeEventListener('beforeunload', this.rescueEditorContentBound);
+  }
+
+  private handleVisibilityChange() {
+    if (document.hidden) {
+      this.stopPolling();
+    } else {
+      void this.updateActivitiesList();
+      this.startPolling();
+    }
+  }
+
+  private startPolling() {
+    if (this.intervallId) window.clearInterval(this.intervallId);
+    this.intervallId = this.pollForUpdates();
+  }
+
+  private stopPolling() {
+    window.clearInterval(this.intervallId);
+  }
+
+  private pollForUpdates() {
+    return window.setInterval(() => this.updateActivitiesList(), this.pollingIntervalInMsValue);
+  }
+
+  handleWorkPackageUpdate(_event:Event):void {
+    setTimeout(() => this.updateActivitiesList(), 2000);
+  }
+
+  async updateActivitiesList() {
+    const journalsContainerAtBottom = this.isJournalsContainerScrolledToBottom(this.journalsContainerTarget);
+    const url = new URL(this.updateStreamsUrlValue);
+    url.searchParams.append('last_update_timestamp', this.lastUpdateTimestamp);
+    url.searchParams.append('filter', this.filterValue);
+
+    const response = await this.fetchWithCSRF(url, 'GET');
+
+    if (response.ok) {
+      const text = await response.text();
+      Turbo.renderStreamMessage(text);
+      this.setLastUpdateTimestamp();
+      setTimeout(() => {
+        if (this.sortingValue === 'asc' && journalsContainerAtBottom) {
+          // scroll to (new) bottom if sorting is ascending and journals container was already at bottom before a new activity was added
+          if (this.isMobile()) {
+            this.scrollInputContainerIntoView(300);
+          } else {
+            this.scrollJournalContainer(this.journalsContainerTarget, true);
+          }
+        }
+      }, 100);
```

Comment:
This delay is rather arbitrary and causing a noticeable delay, also the scrolling duration is very long in my opinion. Please double check with product

Reply:
The timeout is required in order to wait for `Turbo.renderStreamMessage(text)` to have processed and rendered the received stream updates on the actual DOM. `Turbo.renderStreamMessage(text)` does not return a promise or such as far as I investigated. One could use the connect callback of the added item(s) in order to identify when they have been added to the DOM and then trigger the scroll. But this would increase the code complexity quite a bit IMO. Thus I chose to use a timeout of 100ms which should be enough for almost all devices to apply the DOM changes. Additionally I did not see that a delay was noticeable.